### ISSUE TITLE
Comment / uncomment a line (or a selection) with a single shortcut

### ISSUE THREAD

**Issue by RistoVaaraniemi** [wbsoft/frescobaldi#834]
Hi,

I tried to add some comments in my file but I found the process a bit cumbersome (two sequential shortcuts). In some editors (e.g. Notepad++) you can comment / uncomment the whole line with a single shortcut even if the cursor isn't in the start of the line. You can even do that for a selection—if you have selected at least two lines even partially the whole lines are commented out / uncommented. In Notepad++ the command is `Ctrl+Q`.

Would it be feasible to support that in Frescobaldi?

BR,
Risto

**Comment by wbsoft** [comment]
There is a comment and an uncomment snippet, bound to Ctrl+Alt+C, C and Ctrl+Alt+C, U respectively. They work on the cursor position the cursor is. We could change that of course, but I otfen select a part of a line and then comment it, `%{` and `%}` get added. You could modify the snippet to alter the behaviour, to comment/uncomment whole lines. You can also change the keyboard shortcut.

In many programs (also Frescobaldi), Ctrl+Q is connected to Quit. 

Maybe a good solution is to write a Toggle Comment snippet and give that a simple/single shortcut.

**Comment by RistoVaaraniemi** [comment]
OK. Thanks for the pointer. I fiddled a bit and I came up with a Snippet. It uses line based comments to comment / uncomment the selected lines or the line where the cursor is. The lines don't need to be fully selected. If there are lines that are commented and lines that aren't it toggles their states.

For me `Crtl+Q` is convenient because it's easy to reach and I do commenting more often than quitting the app. But every man to his taste ... It can be easily configured—thanks a lot for that. :smiley:

If you'd like to have a look at it, let me know (also how I should deliver it).

BR,
Risto

**Comment by PeteCrighton** [comment]
@RistoVaaraniemi I’d be very interested in that snippet if you’d like to share it.

**Comment by RistoVaaraniemi** [comment]
Hi Pete,

Sorry for noticing your message until now. I tried to attach the snippet as xml, txt and zip but everything failed. I pasted the snippet body in the end of this message, instead. This snippet understands only Lilypond code. I rarely edit scheme, html etc. so I did not bother to support them.

-Risto

-----
-*- python; indent: no;
import re

def main():
    origPosition = cursor.position()
    origAnchor = cursor.anchor()
    origSelectionStart = cursor.selectionStart()
    origSelectionEnd = cursor.selectionEnd()
    
    cursor.setPosition(origSelectionStart)
    cursor.movePosition(cursor.StartOfLine)
    selectionStart = cursor.position()
    
    cursor.setPosition(origSelectionEnd)
    cursor.movePosition(cursor.EndOfLine)
    currentPosition = cursor.position()
    
    lengthCorrection = 0
    firstLineCorrection = 0
    while currentPosition &gt;= selectionStart:
        cursor.select(cursor.LineUnderCursor)
        text = cursor.selection().toPlainText()
        match = re.match(r'^(\s*)%+ ?', text)
        if match == None:
            text = re.sub(r'^(\s*)', r'\1% ', text)
            lengthCorrection += 2
            firstLineCorrection = 2
        else:
            text = re.sub(r'^(\s*)%+ ?', r'\1', text)
            lengthCorrection -= 2
            firstLineCorrection = -2
        
        cursor.removeSelectedText()
        cursor.insertText(text)
        if cursor.movePosition(cursor.Up):
            currentPosition = cursor.position()
        else:
            break
    newSelectionStart = origSelectionStart + firstLineCorrection
    newSelectionEnd = origSelectionEnd + lengthCorrection
    if origAnchor &lt; origPosition:
        cursor.setPosition(newSelectionStart)
        cursor.setPosition(newSelectionEnd, cursor.KeepAnchor)
    else:
        cursor.setPosition(newSelectionEnd)
        cursor.setPosition(newSelectionStart, cursor.KeepAnchor)
    return cursor

**Comment by uliska** [comment]
I agree that it would be nice to have commands to comment/uncomment lines, which should also try to identify the language to be used.
